### PR TITLE
[Small, Important Change] Changed phi_ao code to support puream basis sets

### PR DESCRIPTION
**Pull request by andyj10224**
## Description
Currently, the phi_ao function returns the value of the basis functions at a point. However, this ONLY works for cartesian and not spherical basis sets. I updated the code to make this change. This is important to the development of seminumerical methods, like chain-of-spheres exchange.

## Todos

- [x] Code is tested in my developmental chain-of-spheres exchange branch
- [x] Made an explicit test testing the validity of my bug fix

## Questions
- [x] Will this impact existing functions that use the phi_ao call?

## Status
- [x] Ready for review
- [ ] Ready for merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
Thanks for the diagnosis and the fix.

Why not combine these two functions into a single function with an `if` clause to determine whether the `cart_buffer` transform is necessary?

I'm also going to request you include a test case in _this_ PR.

**Comment by JonathonMisiewicz** [comment]
Following up on an issue from the conference call:
The fixed `phi_ao` functions are used to form the core Hamiltonian whenever an embedding potential is read from a file. While I haven't run anything, I would expect the bug fixed by this PR to cause correctness errors in SCF energies. Definitely agree with getting this fix in for 1.4.

**Comment by andyj10224** [comment]
> Following up on an issue from the conference call:
> The fixed `phi_ao` functions are used to form the core Hamiltonian whenever an embedding potential is read from a file. While I haven't run anything, I would expect the bug fixed by this PR to cause correctness errors in SCF energies. Definitely agree with getting this fix in for 1.4.

Hi Jonathon,

As requested, I cleaned up the code by removing the helper functions, as well as added a new test case (C test phi-ao).
I think having phi_ao as a python function will be useful to Psi developers and users, as it would be immensely helpful for the development of grid-based approaches.

Thank you,
Andy

File: tests/phi-ao/input.dat [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#! Test of the new phi_ao code (for puream and non-puream basis functions)
```
Comment (JonathonMisiewicz) [inline tests/phi-ao/input.dat:1]:
```suggestion
#! Test computing values of basis functions (puream and non-puream) at points
```

File: psi4/src/export_mints.cc [hunk 2]
```diff
@@ -1171,7 +1171,13 @@ void export_mints(py::module& m) {
         .def("max_function_per_shell", &BasisSet::max_function_per_shell,
              "The max number of basis functions in a shell")
         .def("max_nprimitive", &BasisSet::max_nprimitive, "The max number of primitives in a shell")
-        .def_static("construct_from_pydict", &construct_basisset_from_pydict, "docstring");
+        .def_static("construct_from_pydict", &construct_basisset_from_pydict, "docstring")
+        .def("compute_phi", [](BasisSet& basis, double x, double y, double z) {
+            auto phi_ao = new std::vector<double>(basis.nbf());
+            auto capsule = py::capsule(phi_ao, [](void *phi_ao) { delete reinterpret_cast<std::vector<double>*>(phi_ao); });
+            basis.compute_phi(phi_ao->data(), x, y, z);
+            return py::array(phi_ao->size(), phi_ao->data(), capsule);
```
Comment (JonathonMisiewicz) [inline psi4/src/export_mints.cc:1179]: Could you give me a reference for what you're doing here? I can't find one in the Pybind docs and I've never come across code like this before me. I'm guessing this is about [ensuring the data is deleted when appropriate](https://stackoverflow.com/questions/44659924/returning-numpy-arrays-via-pybind11/44682603#44682603)? If so, this is probably going to be very useful.

File: psi4/src/psi4/libmints/basisset.cc [hunk 3]
```diff
@@ -1260,12 +1260,32 @@ void BasisSet::compute_phi(double *phi_ao, double x, double y, double z) {
         double cexpr = 0;
         for (int np = 0; np < nprim; np++) cexpr += c[np] * exp(-a[np] * rr);
 
-        for (int l = 0; l < INT_NCART(am); l++) {
-            Vector3 &components = exp_ao[am][l];
-            phi_ao[ao + l] += pow(dx, (double)components[0]) * pow(dy, (double)components[1]) *
-                              pow(dz, (double)components[2]) * cexpr;
+        if (puream_) {
+            auto s_transform = SphericalTransform(am);
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libmints/basisset.cc:1264]: Does this code compile if you add `const`? I don't believe we've changing this object.

File: psi4/src/psi4/libmints/basisset.cc [hunk 4]
```diff
@@ -1260,12 +1260,32 @@ void BasisSet::compute_phi(double *phi_ao, double x, double y, double z) {
...
+            std::vector<double> cart_buffer(INT_NCART(am), 0.0);
+
+            for (int l = 0; l < INT_NCART(am); l++) {
+                Vector3 &components = exp_ao[am][l];
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libmints/basisset.cc:1268]: `const auto`

File: psi4/src/psi4/libmints/basisset.cc [hunk 5]
```diff
@@ -1260,12 +1260,32 @@ void BasisSet::compute_phi(double *phi_ao, double x, double y, double z) {
...
+                cart_buffer[l] += pow(dx, (double)components[0]) * pow(dy, (double)components[1]) *
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libmints/basisset.cc:1269]: I suggest explicit `static_cast<double>`.

File: psi4/src/psi4/libmints/basisset.cc [hunk 6]
```diff
@@ -1260,12 +1260,32 @@ void BasisSet::compute_phi(double *phi_ao, double x, double y, double z) {
...
+                int lcart = s_transform.cartindex(ind);
+                int lpure = s_transform.pureindex(ind);
+                double coef = s_transform.coef(ind);
+
+                phi_ao[ao + lpure] += coef * cart_buffer[lcart];
+            }
+
+        } else {
+            for (int l = 0; l < INT_NCART(am); l++) {
+                Vector3 &components = exp_ao[am][l];
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libmints/basisset.cc:1283]: Same as previous.

File: tests/phi-ao/input.dat [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        y = np.asarray(block.y())
+        z = np.asarray(block.z())
+
+        for point in range(npoints):
+            phi_ao_points[running_points + point] = basis.compute_phi(x[point], y[point], z[point])
+            weights[running_points + point] = w[point]
+
+        running_points += npoints
+
+    # Creative manipulations for speedy calculations
```
Comment (JonathonMisiewicz) [inline tests/phi-ao/input.dat:54]:
```suggestion
```

File: tests/phi-ao/input.dat [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    basis = psi4.core.BasisSet.build(mol, "ORBITAL", psi4.core.get_global_option("BASIS"), puream=puream)
+    grid = psi4.core.DFTGrid.build(mol, basis)
+
+    # Run an SCF calculation to get Electron Density
+    e, wfn = energy('scf', return_wfn=True)
+
+    # Density Matrix
+    D = np.asarray(wfn.Da())
+    
+    # Grid information
```
Comment (JonathonMisiewicz) [inline tests/phi-ao/input.dat:31]: Move grid setup outside of the for loop. No need to do this twice.

File: tests/phi-ao/input.dat [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  0 1
+  O
+  H 1 1.0
+  H 1 1.0 2 104.5
+  symmetry c1
+  no_reorient
+  no_com
+""")
+
+TEST_NELECTRON = {}
```
Comment (JonathonMisiewicz) [inline tests/phi-ao/input.dat:16]: We don't need this. Just get the value in each for loop and compare it. It doesn't need to persist after that.

**Review by JonathonMisiewicz — COMMENTED** [review]
I think the risk of `compute_phi` changing is minimal, so I'm not opposed to exposing it. "Useful for grid-based approaches" is too vague for me to be _for_ it, but I don't need to be.

**Comment by andyj10224** [comment]
> I think the risk of `compute_phi` changing is minimal, so I'm not opposed to exposing it. "Useful for grid-based approaches" is too vague for me to be _for_ it, but I don't need to be.

An example of "grid-based approaches" would include a psi4numpy implementation of seminumerical methods like chain-of-spheres exchange.

File: psi4/src/export_mints.cc (hunk 2 above)
Comment (andyj10224) [reply psi4/src/export_mints.cc:1179]: Yes, I got this code from this source

https://github.com/pybind/pybind11/issues/1042#issuecomment-325941022

**Comment by andyj10224** [comment]
> I think the risk of `compute_phi` changing is minimal, so I'm not opposed to exposing it. "Useful for grid-based approaches" is too vague for me to be _for_ it, but I don't need to be.

Hi Jonathon, I just made all of your suggested changes. Everything compiled and ran properly :)

**Review by JonathonMisiewicz — APPROVED** [review]

**Review by andysim — APPROVED** [review]
Great test case!  It's a good example of how to use these kinds of objects. LGTM

**Review by loriab — COMMENTED** [review]
looks good, thanks for the nice fix and test! a couple suggestions to consider

File: tests/phi-ao/input.dat [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    phi_prime = np.einsum('uv,pu->pv', D, phi_ao_points)
+    TEST_NELECTRON = 2.0 * np.einsum('p,pv,pv->', weights, phi_prime, phi_ao_points)
+
+    if puream:
+        testname = 'SPHERICAL BASIS PHI_AO TEST'
+    else:
+        testname = 'CARTESIAN BASIS PHI_AO TEST'
+    
+    compare_values(TEST_NELECTRON, REF_NELECTRON, 4, testname)
```
[the remark this answers is not in the document]
Comment (andyj10224) [reply tests/phi-ao/input.dat]: Solved

File: psi4/src/export_mints.cc [hunk 11]
```diff
@@ -1171,7 +1171,13 @@ void export_mints(py::module& m) {
...
+        }, "Calculate the value of all basis functions at a given point x, y, and z");
```
[the remark this answers is not in the document]
Comment (andyj10224) [reply psi4/src/export_mints.cc:1180]: Discussed this with Lori in Slack, but adding the extra arguments led to a compile error.

**Review by loriab — APPROVED** [review]

**Pull request merged**